feat: add beta support to scale Stateful Executors with consensus using RAFT (#5564)
jina-ai/jina: docs/concepts/orchestration/scale-out.md

Flow with three replicas of `slow_encoder` and one replica of `fast_indexer`
```

(scale-consensus)=
## Replicate stateful Executors with consensus using RAFT (Beta)

> ````{admonition} docarray 0.30

Reviewer: ```suggestion
````{admonition} DocArray 0.30
```
Please fix throughout

:class: note

> Since docarray version > 0.30, docarray changed its interface and implementation drastically. We intend to support these new docarray versions
> in the near future, but not every feature is yet available {ref}`Check here <docarray-v2>`. This feature has been added with these new docarray versions support.

Reviewer: ```suggestion
Starting from DocArray version 0.30, DocArray changed its interface and implementation drastically. We intend to support these new versions in the near future, but not every feature is yet available Check {ref}`here <docarray-v2>` for more information. This feature has been added with the new DocArray support.
```

> This feature is only available when using `grpc` as the protocol for the Deployment or when the Deployment is part of a Flow

Reviewer: ```suggestion
This feature is only available when using `grpc` as the protocol for the Deployment or when the Deployment is part of a Flow.
```

Please ensure punctuation throughout

````

````{admonition} gRPC protocol
:class: note

This feature is only available when using gRPC as the protocol for the Deployment or when the Deployment is part of a Flow
````

> Replication is used to scale-out Executors by creating copies of them that can handle requests in parallel providing better RPS.
> However, when an Executor keeps some sort of state, then it is not simple to guarantee that each copy of the Executor keeps the same state,
> which can lead to undesired behavior, since each replica can provide different results, depending on the specific state they hold.

Reviewer: ```suggestion
Replication is used to scale out Executors by creating copies of them that can handle requests in parallel, providing better RPS.
However, when an Executor maintains some sort of state, then it is not simple to guarantee that each copy of the Executor maintains the *same* state,
which can lead to undesired behavior, since each replica can provide different results depending on the specific state they hold.
```

> In Jina, you can also have replication while guaranteeing the consensus between Executors. For this, we rely on (RAFT)[https://raft.github.io/] which is
> an algorithm that guarantees eventual consistency between replicas. 

Reviewer: ```suggestion
In Jina, you can also have replication while guaranteeing the consensus between Executors. For this, we rely on (RAFT)[https://raft.github.io/], which is
an algorithm that guarantees eventual consistency between replicas. 
```

"Eventual" makes it sound like this could take a long time

Author: it is the proper definition of what it does

```python
from jina import Deployment, Executor, requests
from jina.serve.executors.decorators import write
from docarray import DocList
from docarray.documents import TextDoc

class MyStateStatefulExecutor(Executor):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._docs_dict = {}

    @requests(on=['/index'])
    @write
    def index(self, docs: DocList[TextDoc], **kwargs) -> DocList[TextDoc]:
        for doc in docs:
            self._docs_dict[doc.id] = doc

    @requests(on=['/search'])
    def search(self,  docs: DocList[TextDoc], **kwargs) -> DocList[TextDoc]:
        for doc in docs:
            self.logger.debug(f'Searching against {len(self._docs_dict)} documents')
            doc.text = self._docs_dict[doc.id].text

d = Deployment(name='stateful_executor', 
               uses=MyStateStatefulExecutor,
               replicas=3, 
               stateful=True, 
               peer_ports=[12345, 12346, 12347])
with d:
    d.block()
```

This capacity allows us not only to have replicas work with robustness and availability, it also can help us to achieve higher throughput in some cases.

> Let's imagine we write an Executor that is used to index and query documents from a vector index.

Author: as u want, let's do this yes

Flow with three replicas of `slow_encoder` and one replica of `fast_indexer`
```

(scale-consensus)=
## Replicate stateful Executors with consensus using RAFT (Beta)

````{admonition} DocArray 0.30
:class: note

> Starting from DocArray version 0.30, DocArray changed its interface and implementation drastically. We intend to support these new versions in the near future, but not every feature is yet available Check {ref}`here <docarray-v2>` for more information. This feature has been added with the new DocArray support.

Reviewer: ```suggestion
Starting from DocArray version 0.30, DocArray changed its interface and implementation drastically. We intend to support these new versions in the near future, but not every feature is yet available. Check {ref}`here <docarray-v2>` for more information. This feature has been added with the new DocArray support.
```

> This feature is only available when using `grpc` as the protocol for the Deployment or when the Deployment is part of a Flow.

Reviewer: ```suggestion
```
This is duplicated below

an algorithm that guarantees eventual consistency between replicas. 

Consensus-based replication using RAFT is a distributed algorithm designed to provide fault tolerance and consistency in a distributed system. In a distributed system, the nodes may fail, and messages may be lost or delayed, which can lead to inconsistencies in the system.
The problem with traditional replication methods is that they can't guarantee consistency in a distributed system in the presence of failures. This is where consensus-based replication using RAFT comes in.
In this approach, each Executor can be considered as a Finite State Machine, which means that it has a set of states that it can be in, and a set of transitions that it can make between those states. Each request that is sent to the Executor can be considered as a log entry that needs to be replicated across the cluster.

In order to enable this kind of replication, we need to consider:

- Specify which methods of the Executor {ref}` could be updating its internal state <stateful-executor>`.

> - Tell the deployment to use the RAFT consensus algorithm by setting the `--stateful` argument.

Author: ```suggestion
- Tell the Deployment to use the RAFT consensus algorithm by setting the `--stateful` argument.
```

INFO   indexer/rep-0@902 Indexing Document in index with 99700 documents indexed                                                                                                                                   
INFO   indexer/rep-0@902 Indexing Document in index with 99800 documents indexed                                                                                                                                   
INFO   indexer/rep-0@902 Indexing Document in index with 99900 documents indexed 
```

But at search time, the consensus module does not affect, and only one replica serves the queries.
```text
INFO   indexer/rep-0@902 Searching Document in index with 100000 documents indexed                                                                                                              [04/28/23 16:59:21]
INFO   indexer/rep-1@910 Searching Document in index with 100000 documents indexed                                                                                                              [04/28/23 16:59:21]
INFO   indexer/rep-2@923 Searching Document in index with 100000 documents indexed 
```

> If you run the same example by setting replicas to 1 without the consensus module, you can see the benefits it has in the QPS at search time,

Reviewer: ```suggestion
If you run the same example by setting `replicas` to `1` without the consensus module, you can see the benefits it has in the QPS at search time,
```

while there is a little cost on the time used for indexing.

```python
d = Deployment(name='indexer',
               port=5555,
               uses=ExactNNSearch,
               replicas=1)
```

> With 1 replica:

Reviewer: ```suggestion
With one replica:
```

```text
Indexing 100000 Documents took 18.93274688720703s
Searching 1000 Queries took 385.96641397476196s
```

> With 3 replicas and consensus:

Reviewer: ```suggestion
With three replicas and consensus:
```

```text
Indexing 100000 Documents took 35.066415548324585s
Searching 1000 Queries took 202.07950615882874s
```

> This allows to go from 2.5 to 5 QPS.

Reviewer: ```suggestion
This increases QPS from 2.5 to 5.
```
